gef.py

```diff
@@ -2236,18 +2235,18 @@ def get_info_sections():
     return
 
 def get_saved_ip():
-    """Retrieves the saved ip using the 'info frame' command ."""
+    """Retrieves the saved ip using the 'info frame' command."""
 
     lines = gdb.execute("info frame", to_string=True).splitlines()
```

Comment:
Using `gdb.execute` should be avoided as much as possible in favor of pure Python binding (if they exist): please look at https://sourceware.org/gdb/onlinedocs/gdb/Frames-In-Python.html and update your code.

Pure Python bindings are more readable, faster and less prone to change from one GDB to another.

Reply:
Hi, 
Thanks for the response.  

I spent about an hour looking at the gdb Python-API's including the one you suggested.  I could not find a method that provided the address of the saved instruction pointer. 

Did you have something specific in mind?

-Erik

Reply:
Hey!

I was talking about using GDB-Python [`Frame`](https://sourceware.org/gdb/onlinedocs/gdb/Frames-In-Python.html) instead of calling `gdb.execute("info frame")`. The API allows you to navigate through GDB's frames in pure Python (no string parsing), to get (among other things, the frame `name()` and its `pc()` which is exactly what you want.

Take a look at the `context_trace()` function of the `ContextCommand` class, it'll give you a better idea.

Cheers

Reply:
Thank you again for replying,  I'm really sorry, but I'm still  confused even though I spent some more time looking the docs on the API and  how the gef.py uses the gdb Frame object.  Unfortunately, I don't see how the name of the function (.e.,g main) and the current address that's being executed can get me the saved instruction pointer address for the function frame.Have you seen this done somewhere with a Frame object?

If you are thinking it'd be better if I did a calculation instead for each of the architecture supported by gef, I could do that.  But the gdb.execute("info frame") seemed so nice because it worked for all the architectures.  In the alternative, I could cache the value so that it has as little impact to performance as possible.

Any more guidance would be greatly appreciated.

Thanks!

Reply:
From the [docs](https://sourceware.org/gdb/onlinedocs/gdb/Registers.html#Registers):
 > `$fp` is used for a register that contains a pointer to the current stack frame,[...]

Example with `/bin/ls` for instance:
```bash
$ gdb -ex 'bp main' -ex r /bin/ls
```

Then with your current code (simplified)
```python
gef➤  python print(hex(int(gdb.execute("info frame", to_string=True).splitlines()[-1].split()[-1], 16)))
0x7fffffffe3e8
```

Which in pure Python would be something like:
```python
gef➤  python print(hex(gdb.selected_frame().older().read_register('fp')))
0x7fffffffe3e8
```

The latter is smaller, cleaner (no string manipulation), more portable because 100% relying on Python API.